### ISSUE TITLE
Validation inference has no chunking, and the memory budget does not cover it

### ISSUE THREAD

**Issue by matthiaskloft** [matthiaskloft/bayesflow-hpo#101]
Validation inference samples an entire condition batch in one call, with no
chunking and no memory budget covering it. A trial can pass the pre-training
budget check and then OOM during validation.

Found while measuring the GPU cost ratio for #82; it is independent of that
work and reproduces on `main`.

## What the code does

`make_bayesflow_infer_fn` builds the closure the validation pipeline uses, and
that closure makes exactly one call
([`validation/inference.py:53`](https://github.com/matthiaskloft/bayesflow-hpo/blob/main/src/bayesflow_hpo/validation/inference.py)):

```python
post_draws = approximator.sample(
    conditions=conditions, num_samples=int(n_posterior_samples),
)
```

`conditions` is the whole condition batch. `run_validation_pipeline` calls this
once per condition (`validation/pipeline.py:79`), so the peak allocation is
`n_sims × n_posterior_samples` posterior samples materialized at once, plus
every intermediate the inference network's forward pass needs for that many
rows. Nothing chunks, and nothing caps it.

## Why the defaults are already large

- `generate_validation_dataset(sims_per_condition=200)` (`validation/data.py:65`)
- `optimize(n_posterior_samples=500)` (`api.py:126`), and
  `run_validation_pipeline(n_posterior_samples=1000)` (`validation/pipeline.py:26`)

So a default study already asks for **100,000 samples in a single forward
pass**, and the pipeline's own default is 200,000. This is not an exotic
configuration.

## Measured

On an RTX 5090 (32 GiB, driver 615.71.09, torch 2.11.0+cu128, bayesflow 2.0.8),
a FlowMatching + DeepSet approximator with `n_params=15`, `n_obs=50`:

| n_sims × n_draws | samples | result |
| --- | ---: | --- |
| 100 × 200 | 20,000 | ok |
| 200 × 200 | 40,000 | ok |
| 100 × 400 | 40,000 | ok |
| 150 × 400 | 60,000 | OOM at a 6.29 GiB process cap |
| 500 × 1000 | 500,000 | OOM with ~20 GiB free |

The 500 × 1000 case failed inside
`FlowMatching._inverse` → `integrate_adaptive` → `tsit5_step`, on a
`(500000, 256)` activation. An adaptive ODE integrator holds several stages
live at once, so the multiplier on the nominal sample count is substantial.

## Why the existing budget guard does not catch it

`estimate_peak_memory_mb` (`optimization/constraints.py:197`) is documented as
estimating "approximate peak **training** memory" — parameters × 4 for Adam's
buffers plus an activation proxy from `batch_size`. It is evaluated
*pre-training* to reject oversized configs before allocating. Validation
inference is a different allocation entirely: it is driven by
`n_sims × n_posterior_samples`, neither of which is a search-space
hyperparameter and neither of which appears in that estimate.

So `max_memory_mb` (including `"auto"`) can pass a trial that then dies in
validation. The failure also arrives *after* training has been paid for, which
is the most expensive moment to discover it.

## Impact

- A trial that OOMs in validation loses its training work and falls back to the
  training-loss path, so the study records a penalty for what is a resource
  problem rather than a model-quality one.
- It caps how large a validation set can be for a given card, silently and
  with no diagnostic pointing at `sims_per_condition` or
  `n_posterior_samples` as the cause.
- It is worse on GPU than on CPU, which is the configuration the package is
  aimed at.

## Possible directions

Not a design decision yet, but the options look like:

1. **Chunk inside `infer_fn`** — loop over slices of the condition batch and
   concatenate. The metric contract already expects the assembled
   `(n_sims, n_samples, n_params)` array, so this is invisible above
   `inference.py`. Chunk size would need a default and an override.
2. **Extend the budget estimate** to cover validation inference, so a trial is
   rejected pre-training rather than dying post-training.
3. **Catch OOM in validation and retry at a smaller chunk**, which is what a
   caller would otherwise do by hand.

(1) looks like the smallest change that removes the failure; (2) is what makes
the existing `max_memory_mb` honest.

## Measured while the card was shared

The benchmark ran under a deliberate `set_per_process_memory_fraction` cap so
an overrun would fail the benchmark rather than an unrelated job on the same
card — the 6.29 GiB figure above is that self-cap, not the card's limit. The
500 × 1000 case is the one that exceeded genuinely free memory.

Harness: `docs/plans/bench_inference_ratio.py` on branch
`claude/issue-82-37fb3c`.

Related: #82 (found during its cost measurement; that plan's D10 is corrected
to point here).

— Claude Code (Claude Opus 5)
